File: app/src/main/java/com/expansion/lg/kimaru/expansion/other/ExportDataToCsv.java
package com.expansion.lg.kimaru.expansion.other;

import android.content.Context;
import android.os.Environment;
import android.text.TextUtils;
import android.widget.Toast;

import com.expansion.lg.kimaru.expansion.mzigos.ChewReferral;
import com.expansion.lg.kimaru.expansion.mzigos.CommunityUnit;
import com.expansion.lg.kimaru.expansion.mzigos.Mapping;
import com.expansion.lg.kimaru.expansion.mzigos.Mobilization;
import com.expansion.lg.kimaru.expansion.mzigos.Parish;
import com.expansion.lg.kimaru.expansion.mzigos.Village;
import com.expansion.lg.kimaru.expansion.tables.ChewReferralTable;
import com.expansion.lg.kimaru.expansion.tables.CommunityUnitTable;
import com.expansion.lg.kimaru.expansion.tables.MappingTable;
import com.expansion.lg.kimaru.expansion.tables.MobilizationTable;
import com.expansion.lg.kimaru.expansion.tables.ParishTable;
import com.expansion.lg.kimaru.expansion.tables.UserTable;
import com.expansion.lg.kimaru.expansion.tables.VillageTable;

import java.io.File;
import java.io.FileWriter;
import java.io.PrintWriter;
import java.util.HashMap;
import java.util.List;

/**
 * Created by kimaru on 8/22/17.
 */

public class ExportDataToCsv {
    Context context;
    String state;

    public ExportDataToCsv(Context context) {
        this.context = context;
        state = Environment.getExternalStorageState();
    }

    public void exportChewReferral(){
        ChewReferralTable chewReferralTable = new ChewReferralTable(context);
        List<ChewReferral> chewReferrals = chewReferralTable.getChewReferralData();
        String[] headers = chewReferralTable.columns;
        if (!Environment.MEDIA_MOUNTED.equals(state)){
            Toast.makeText(context, "Media is not mounted.", Toast.LENGTH_SHORT).show();
            return;
        }else{
            PrintWriter printWriter = exportCsv("chew_referrals");
            String csvHeader = TextUtils.join(",", headers);
            printWriter.println(csvHeader);
            for (ChewReferral p:chewReferrals){
                String detailsLine = p.getId()+","
                        + p.getName().replaceAll(",", ";")+","
                        + p.getPhone().replaceAll(",", ";")+","
                        + p.getTitle().replaceAll(",", ";")+","
                        + p.getCountry().replaceAll(",", ";")+","
                        + p.getRecruitmentId().replaceAll(",", ";")+","
                        + p.getSynced()+","
                        + p.getCounty().replaceAll(",", ";")+","
                        + p.getDistrict().replaceAll(",", ";")+","
                        + p.getSubCounty().replaceAll(",", ";")+","
                        + p.getCommunityUnit().replaceAll(",", ";")+","
                        + p.getVillage().replaceAll(",", ";")+","
                        + p.getMapping().replaceAll(",", ";")+","
                        + p.getMobilization().replaceAll(",", ";")+","
                        + p.getLon().replaceAll(",", ";")+","
                        + p.getLat().replaceAll(",", ";");
                printWriter.println(detailsLine);
            }
            if(printWriter != null) printWriter.close();

        }
    }


    public void exportCommunityUnit(){
        CommunityUnitTable communityUnitTable = new CommunityUnitTable(context);
        List<CommunityUnit> communityUnits = communityUnitTable.getCommunityUnitData();
        String[] headers = communityUnitTable.columns;
        if (!Environment.MEDIA_MOUNTED.equals(state)){
            Toast.makeText(context, "Media is not mounted.", Toast.LENGTH_SHORT).show();
            return;
        }else{
            PrintWriter printWriter = exportCsv("community_units");
            String csvHeader = TextUtils.join(",", headers);
            printWriter.println(csvHeader);
            for (CommunityUnit cm:communityUnits){
                String detailsLine = cm.getId()+","
                        + (cm.getCommunityUnitName() == null ? "" : cm.getCommunityUnitName().replaceAll(",", ";"))+","
                        + (cm.getMappingId() == null ? "" : cm.getMappingId().replaceAll(",", ";"))+","
                        + (cm.getLat() == null ? "" : cm.getLat().toString())+","
                        + (cm.getLon() == null ? "" : cm.getLon().toString())+","
                        + (cm.getCountry() == null ? "" : cm.getCountry().replaceAll(",",";"))+","
                        + (cm.getSubCountyId() == null ? "" : cm.getSubCountyId().replaceAll(",",";"))+","
                        + (cm.getLinkFacilityId() == null ? "" : cm.getLinkFacilityId().replaceAll(",",";"))+","
                        + (cm.getAreaChiefName() == null ? "" : cm.getAreaChiefName().replaceAll(",",";"))+","
                        + (cm.getAreaChiefPhone() == null ? "" : cm.getAreaChiefPhone().replaceAll(",",";"))+","
                        + (cm.getWard() == null ? "" : cm.getWard().replaceAll(",",";"))+","
                        + (cm.getEconomicStatus() == null ? "" : cm.getEconomicStatus().replaceAll(",",";"))+","
                        + (cm.getPrivateFacilityForAct() == null ? "" : cm.getPrivateFacilityForAct().replaceAll(",",";"))+","
                        + (cm.getPrivateFacilityForMrdt() == null ? "" : cm.getPrivateFacilityForMrdt().replaceAll(",",";"))+","
                        + (cm.getNameOfNgoDoingIccm() == null ? "" : cm.getNameOfNgoDoingIccm().replaceAll(",",";"))+","
                        + (cm.getNameOfNgoDoingMhealth() == null ? "" : cm.getNameOfNgoDoingMhealth().replaceAll(",",";"))+","
                        + cm.getDateAdded()+","
                        + new UserTable(context).getUserById(Integer.valueOf(String.valueOf(cm.getAddedBy()))).getName()+","
                        + cm.getNumberOfChvs()+","
                        + cm.getHouseholdPerChv()+","
                        + cm.getNumberOfVillages()+","
                        + cm.getDistanceToBranch()+","
                        + cm.getTransportCost()+","
                        + cm.getDistanceTOMainRoad()+","
                        + cm.getNoOfHouseholds()+","
                        + cm.getMohPoplationDensity()+","
                        + cm.getEstimatedPopulationDensity()+","
                        + cm.getDistanceTONearestHealthFacility()+","
                        + cm.getActLevels()+","
                        + cm.getActPrice()+","
                        + cm.getMrdtLevels()+","
                        + cm.getMrdtPrice()+","
                        + cm.getNoOfDistibutors()+","
                        + cm.isChvsTrained()+","
                        + cm.isPresenceOfEstates()+","
                        + cm.isPresenceOfFactories()+","
                        + cm.isPresenceOfHostels()+","
                        + cm.isTraderMarket()+","
                        + cm.isLargeSupermarket()+","
                        + cm.isNgosGivingFreeDrugs()+","
                        + cm.isNgoDoingIccm()+","
                        + cm.isNgoDoingMhealth();
                printWriter.println(detailsLine);
            }
            if(printWriter != null) printWriter.close();

        }
    }


    public void exportMobilization(){
        MobilizationTable mobilizationTable = new MobilizationTable(context);
        List<Mobilization> mobilizations = mobilizationTable.getMobilizationData();
        String[] headers = mobilizationTable.columns;
        if (!Environment.MEDIA_MOUNTED.equals(state)){
            Toast.makeText(context, "Media is not mounted.", Toast.LENGTH_SHORT).show();
            return;
        }else{
            PrintWriter printWriter = exportCsv("mobilization");
            String csvHeader = TextUtils.join(",", headers);
            printWriter.println(csvHeader);
            for (Mobilization m:mobilizations){
                String detailsLine = m.getId()+","
                        + m.getName().replaceAll(",", ";")+","
                        + m.getMappingId().replaceAll(",", ";")+","
                        + m.getCountry().replaceAll(",", ";")+","
                        + new UserTable(context).getUserById(m.getAddedBy()).getName()+","
                        + m.getComment().replaceAll(",", ";")+","
                        + m.getDateAdded()+","
                        + m.getSynced()+","
                        + m.getDistrict().replaceAll(",", ";")+","
                        + m.getCounty().replaceAll(",", ";")+","
                        + m.getSubCounty().replaceAll(",", ";")+","
                        + m.getParish().replaceAll(",", ";");
                printWriter.println(detailsLine);
            }
            if(printWriter != null) printWriter.close();

        }
    }


    public void exportVillage(){
        VillageTable villageTable = new VillageTable(context);
        List<Village> villages = villageTable.getVillageData();
        String[] headers = villageTable.columns;
        if (!Environment.MEDIA_MOUNTED.equals(state)){
            Toast.makeText(context, "Media is not mounted.", Toast.LENGTH_SHORT).show();
            return;
        }else{
            PrintWriter printWriter = exportCsv("villages");
            String csvHeader = TextUtils.join(",", headers);
            printWriter.println(csvHeader);
            for (Village m:villages){
                String detailsLine = m.getId()+","
                        + m.getVillageName().replaceAll(",", ";")+","
                        + m.getMappingId().replaceAll(",", ";")+","
                        + m.getLat()+","
                        + m.getLon()+","
                        + m.getCountry().replaceAll(",", ";")+","
                        + m.getDistrict().replaceAll(",", ";")+","
                        + m.getCounty().replaceAll(",", ";")+","
                        + m.getSubCountyId().replaceAll(",", ";")+","
                        + m.getParish().replaceAll(",", ";")+","
                        + m.getCommunityUnit().replaceAll(",", ";")+","
                        + m.getWard().replaceAll(",", ";")+","
                        + m.getLinkFacilityId().replaceAll(",", ";")+","
                        + m.getAreaChiefName().replaceAll(",", ";")+","
                        + m.getAreaChiefPhone().replaceAll(",", ";")+","
                        + m.getDistanceToBranch()+","
                        + m.getTransportCost()+","
                        + m.getDistanceToMainRoad()+","
                        + m.getNoOfHouseholds()+","
                        + m.getMohPoplationDensity()+","
                        + m.getEstimatedPopulationDensity()+","
                        + m.getEconomicStatus()+","
                        + m.getDistanceToNearestHealthFacility()+","
                        + m.getActLevels()+","
                        + m.getActPrice()+","
                        + m.getMrdtLevels()+","
                        + m.getMrdtPrice()+","
                        + m.getPresenceOfHostels()+","
                        + m.getPresenceOfEstates()+","
                        + m.getNumberOfFactories()+","
                        + m.getPresenceOfDistributors()+","
                        + m.getDistributorsInTheArea()+","
                        + m.getTraderMarket()+","
                        + m.getLargeSupermarket()+","
                        + m.getNgosGivingFreeDrugs()+","
                        + m.getNgoDoingIccm()+","
                        + m.getNgoDoingMhealth()+","
                        + m.getNameOfNgoDoingIccm().replaceAll(",",";")+","
                        + m.getNameOfNgoDoingMhealth().replaceAll(",",";")+","
                        + m.getPrivateFacilityForAct().replaceAll(",",";")+","
                        + m.getPrivateFacilityForMrdt().replaceAll(",",";")+","
                        + m.getDateAdded()+","
                        + m.getComment().replaceAll(",",";")+","
                        + m.isSynced()+","
                        + m.getChvsTrained()+","
                        + m.isBracOperating()+","
                        + m.getSafaricomSignalStrength()+","
                        + m.getMtnSignalStrength()+","
                        + m.getAirtelSignalStrength()+","
                        + m.getOrangeSignalStrength()+","
                        + m.getActStock();
                printWriter.println(detailsLine);
            }
            if(printWriter != null) printWriter.close();
        }

    }
    public void exportParish(){
        ParishTable parishTable = new ParishTable(context);
        List<Parish> parishes = parishTable.getParishData();
        String[] headers = parishTable.columns;
        if (!Environment.MEDIA_MOUNTED.equals(state)){
            Toast.makeText(context, "Media is not mounted.", Toast.LENGTH_SHORT).show();
            return;
        }else{
            PrintWriter printWriter = exportCsv("parish");
            String csvHeader = TextUtils.join(",", headers);
            printWriter.println(csvHeader);
            for (Parish p:parishes){
                String detailsLine = p.getId()+","
                        + p.getName().replaceAll(",", ";")+","
                        + p.getCountry().replaceAll(",", ";")+","
                        + p.getParent().replaceAll(",", ";")+","
                        + p.getMapping().replaceAll(",", ";")+","
                        + new UserTable(context).getUserById(p.getAddedBy()).getName()+","
                        + p.getContactPerson().replaceAll(",", ";")+","
                        + p.getContactPersonPhone().replaceAll(",", ";")+","
                        + p.getComment().replaceAll(",", ";")+","
                        + p.getSynced()+","
                        + p.getDateAdded();
                printWriter.println(detailsLine);
            }
            if(printWriter != null) printWriter.close();

        }
    }
    public void exportMapping(){
        MappingTable mappingTable = new MappingTable(context);
        List<Mapping> mappings = mappingTable.getMappingData();
        String[] headers = mappingTable.columns;
        if (!Environment.MEDIA_MOUNTED.equals(state)){
            Toast.makeText(context, "Media is not mounted.", Toast.LENGTH_SHORT).show();
            return;
        }else{
            PrintWriter printWriter = exportCsv("mapping");
            String csvHeader = TextUtils.join(",", headers);
            printWriter.println(csvHeader);
            for (Mapping m:mappings){
                String detailsLine = m.getId()+","
                        + m.getMappingName().replaceAll(",", ";")+","
                        + m.getCountry().replaceAll(",", ";")+","
                        + m.getCounty().replaceAll(",", ";")+","
                        + new UserTable(context).getUserById(m.getAddedBy()).getName()+","
                        + m.getContactPerson().replaceAll(",", ";")+","
                        + m.getContactPersonPhone().replaceAll(",", ";")+","
                        + m.getComment().replaceAll(",", ";")+","
                        + m.getDateAdded()+","
                        + m.isSynced()+","
                        + m.getDistrict().replaceAll(",", ";")+","
                        + m.getSubCounty().replaceAll(",", ";");
                printWriter.println(detailsLine);
            }
            if(printWriter != null) printWriter.close();

        }

    }

    private PrintWriter exportCsv(String dataToExport){
        File exportDir = new File(Environment.getExternalStorageDirectory() + "/Download/tremap/");
        if (!exportDir.exists()){
            exportDir.mkdirs();
        }
        File file;
        PrintWriter printWriter = null;
        file = new File(exportDir, dataToExport+".csv");
        try {
            file.createNewFile();
            printWriter = new PrintWriter(new FileWriter(file));

        }catch(Exception e){}
        return printWriter;
    }
}

File: app/src/main/java/com/expansion/lg/kimaru/expansion/other/ExportDataToCsv.kt
package com.expansion.lg.kimaru.expansion.other

import android.content.Context
import android.os.Environment
import android.text.TextUtils
import android.widget.Toast

import com.expansion.lg.kimaru.expansion.mzigos.ChewReferral
import com.expansion.lg.kimaru.expansion.mzigos.CommunityUnit
import com.expansion.lg.kimaru.expansion.mzigos.Mapping
import com.expansion.lg.kimaru.expansion.mzigos.Mobilization
import com.expansion.lg.kimaru.expansion.mzigos.Parish
import com.expansion.lg.kimaru.expansion.mzigos.Village
import com.expansion.lg.kimaru.expansion.tables.ChewReferralTable
import com.expansion.lg.kimaru.expansion.tables.CommunityUnitTable
import com.expansion.lg.kimaru.expansion.tables.MappingTable
import com.expansion.lg.kimaru.expansion.tables.MobilizationTable
import com.expansion.lg.kimaru.expansion.tables.ParishTable
import com.expansion.lg.kimaru.expansion.tables.UserTable
import com.expansion.lg.kimaru.expansion.tables.VillageTable

import java.io.File
import java.io.FileWriter
import java.io.PrintWriter
import java.util.HashMap

/**
 * Created by kimaru on 8/22/17.
 */

class ExportDataToCsv(internal var context: Context) {
    internal var state: String

    init {
        state = Environment.getExternalStorageState()
    }

    fun exportChewReferral() {
        val chewReferralTable = ChewReferralTable(context)
        val chewReferrals = chewReferralTable.chewReferralData
        val headers = chewReferralTable.columns
        if (Environment.MEDIA_MOUNTED != state) {
            Toast.makeText(context, "Media is not mounted.", Toast.LENGTH_SHORT).show()
            return
        } else {
            val printWriter = exportCsv("chew_referrals")
            val csvHeader = TextUtils.join(",", headers)
            printWriter!!.println(csvHeader)
            for (p in chewReferrals) {
                val detailsLine = (p.id + ","
                        + p.name.replace(",".toRegex(), ";") + ","
                        + p.phone.replace(",".toRegex(), ";") + ","
                        + p.title.replace(",".toRegex(), ";") + ","
                        + p.country.replace(",".toRegex(), ";") + ","
                        + p.recruitmentId.replace(",".toRegex(), ";") + ","
                        + p.synced + ","
                        + p.county.replace(",".toRegex(), ";") + ","
                        + p.district.replace(",".toRegex(), ";") + ","
                        + p.subCounty.replace(",".toRegex(), ";") + ","
                        + p.communityUnit.replace(",".toRegex(), ";") + ","
                        + p.village.replace(",".toRegex(), ";") + ","
                        + p.mapping.replace(",".toRegex(), ";") + ","
                        + p.mobilization.replace(",".toRegex(), ";") + ","
                        + p.lon.replace(",".toRegex(), ";") + ","
                        + p.lat.replace(",".toRegex(), ";"))
                printWriter.println(detailsLine)
            }
            printWriter?.close()

        }
    }


    fun exportCommunityUnit() {
        val communityUnitTable = CommunityUnitTable(context)
        val communityUnits = communityUnitTable.communityUnitData
        val headers = communityUnitTable.columns
        if (Environment.MEDIA_MOUNTED != state) {
            Toast.makeText(context, "Media is not mounted.", Toast.LENGTH_SHORT).show()
            return
        } else {
            val printWriter = exportCsv("community_units")
            val csvHeader = TextUtils.join(",", headers)
            printWriter!!.println(csvHeader)
            for (cm in communityUnits) {
                val detailsLine = (cm.id + ","
                        + (if (cm.communityUnitName == null) "" else cm.communityUnitName.replace(",".toRegex(), ";")) + ","
                        + (if (cm.mappingId == null) "" else cm.mappingId.replace(",".toRegex(), ";")) + ","
                        + (if (cm.lat == null) "" else cm.lat!!.toString()) + ","
                        + (if (cm.lon == null) "" else cm.lon!!.toString()) + ","
                        + (if (cm.country == null) "" else cm.country.replace(",".toRegex(), ";")) + ","
                        + (if (cm.subCountyId == null) "" else cm.subCountyId.replace(",".toRegex(), ";")) + ","
                        + (if (cm.linkFacilityId == null) "" else cm.linkFacilityId.replace(",".toRegex(), ";")) + ","
                        + (if (cm.areaChiefName == null) "" else cm.areaChiefName.replace(",".toRegex(), ";")) + ","
                        + (if (cm.areaChiefPhone == null) "" else cm.areaChiefPhone.replace(",".toRegex(), ";")) + ","
                        + (if (cm.ward == null) "" else cm.ward.replace(",".toRegex(), ";")) + ","
                        + (if (cm.economicStatus == null) "" else cm.economicStatus.replace(",".toRegex(), ";")) + ","
                        + (if (cm.privateFacilityForAct == null) "" else cm.privateFacilityForAct.replace(",".toRegex(), ";")) + ","
                        + (if (cm.privateFacilityForMrdt == null) "" else cm.privateFacilityForMrdt.replace(",".toRegex(), ";")) + ","
                        + (if (cm.nameOfNgoDoingIccm == null) "" else cm.nameOfNgoDoingIccm.replace(",".toRegex(), ";")) + ","
                        + (if (cm.nameOfNgoDoingMhealth == null) "" else cm.nameOfNgoDoingMhealth.replace(",".toRegex(), ";")) + ","
                        + cm.dateAdded + ","
                        + UserTable(context).getUserById(Integer.valueOf(cm.addedBy.toString()))!!.name + ","
                        + cm.numberOfChvs + ","
                        + cm.householdPerChv + ","
                        + cm.numberOfVillages + ","
                        + cm.distanceToBranch + ","
                        + cm.transportCost + ","
                        + cm.distanceTOMainRoad + ","
                        + cm.noOfHouseholds + ","
                        + cm.mohPoplationDensity + ","
                        + cm.estimatedPopulationDensity + ","
                        + cm.distanceTONearestHealthFacility + ","
                        + cm.actLevels + ","
                        + cm.actPrice + ","
                        + cm.mrdtLevels + ","
                        + cm.mrdtPrice + ","
                        + cm.noOfDistibutors + ","
                        + cm.isChvsTrained + ","
                        + cm.isPresenceOfEstates + ","
                        + cm.isPresenceOfFactories + ","
                        + cm.isPresenceOfHostels + ","
                        + cm.isTraderMarket + ","
                        + cm.isLargeSupermarket + ","
                        + cm.isNgosGivingFreeDrugs + ","
                        + cm.isNgoDoingIccm + ","
                        + cm.isNgoDoingMhealth)
                printWriter.println(detailsLine)
            }
            printWriter?.close()

        }
    }


    fun exportMobilization() {
        val mobilizationTable = MobilizationTable(context)
        val mobilizations = mobilizationTable.mobilizationData
        val headers = mobilizationTable.columns
        if (Environment.MEDIA_MOUNTED != state) {
            Toast.makeText(context, "Media is not mounted.", Toast.LENGTH_SHORT).show()
            return
        } else {
            val printWriter = exportCsv("mobilization")
            val csvHeader = TextUtils.join(",", headers)
            printWriter!!.println(csvHeader)
            for (m in mobilizations) {
                val detailsLine = (m.id + ","
                        + m.name.replace(",".toRegex(), ";") + ","
                        + m.mappingId.replace(",".toRegex(), ";") + ","
                        + m.country.replace(",".toRegex(), ";") + ","
                        + UserTable(context).getUserById(m.addedBy!!)!!.name + ","
                        + m.comment.replace(",".toRegex(), ";") + ","
                        + m.dateAdded + ","
                        + m.synced + ","
                        + m.district.replace(",".toRegex(), ";") + ","
                        + m.county.replace(",".toRegex(), ";") + ","
                        + m.subCounty.replace(",".toRegex(), ";") + ","
                        + m.parish.replace(",".toRegex(), ";"))
                printWriter.println(detailsLine)
            }
            printWriter?.close()

        }
    }


    fun exportVillage() {
        val villageTable = VillageTable(context)
        val villages = villageTable.villageData
        val headers = VillageTable.columns
        if (Environment.MEDIA_MOUNTED != state) {
            Toast.makeText(context, "Media is not mounted.", Toast.LENGTH_SHORT).show()
            return
        } else {
            val printWriter = exportCsv("villages")
            val csvHeader = TextUtils.join(",", headers)
            printWriter!!.println(csvHeader)
            for (m in villages) {
                val detailsLine = (m.id + ","
                        + m.villageName.replace(",".toRegex(), ";") + ","
                        + m.mappingId.replace(",".toRegex(), ";") + ","
                        + m.lat + ","
                        + m.lon + ","
                        + m.country.replace(",".toRegex(), ";") + ","
                        + m.district.replace(",".toRegex(), ";") + ","
                        + m.county.replace(",".toRegex(), ";") + ","
                        + m.subCountyId.replace(",".toRegex(), ";") + ","
                        + m.parish.replace(",".toRegex(), ";") + ","
                        + m.communityUnit.replace(",".toRegex(), ";") + ","
                        + m.ward.replace(",".toRegex(), ";") + ","
                        + m.linkFacilityId.replace(",".toRegex(), ";") + ","
                        + m.areaChiefName.replace(",".toRegex(), ";") + ","
                        + m.areaChiefPhone.replace(",".toRegex(), ";") + ","
                        + m.distanceToBranch + ","
                        + m.transportCost + ","
                        + m.distanceToMainRoad + ","
                        + m.noOfHouseholds + ","
                        + m.mohPoplationDensity + ","
                        + m.estimatedPopulationDensity + ","
                        + m.economicStatus + ","
                        + m.distanceToNearestHealthFacility + ","
                        + m.actLevels + ","
                        + m.actPrice + ","
                        + m.mrdtLevels + ","
                        + m.mrdtPrice + ","
                        + m.presenceOfHostels + ","
                        + m.presenceOfEstates + ","
                        + m.numberOfFactories + ","
                        + m.presenceOfDistributors + ","
                        + m.distributorsInTheArea + ","
                        + m.traderMarket + ","
                        + m.largeSupermarket + ","
                        + m.ngosGivingFreeDrugs + ","
                        + m.ngoDoingIccm + ","
                        + m.ngoDoingMhealth + ","
                        + m.nameOfNgoDoingIccm.replace(",".toRegex(), ";") + ","
                        + m.nameOfNgoDoingMhealth.replace(",".toRegex(), ";") + ","
                        + m.privateFacilityForAct.replace(",".toRegex(), ";") + ","
                        + m.privateFacilityForMrdt.replace(",".toRegex(), ";") + ","
                        + m.dateAdded + ","
                        + m.comment.replace(",".toRegex(), ";") + ","
                        + m.isSynced + ","
                        + m.chvsTrained + ","
                        + m.isBracOperating + ","
                        + m.safaricomSignalStrength + ","
                        + m.mtnSignalStrength + ","
                        + m.airtelSignalStrength + ","
                        + m.orangeSignalStrength + ","
                        + m.actStock)
                printWriter.println(detailsLine)
            }
            printWriter?.close()
        }

    }

    fun exportParish() {
        val parishTable = ParishTable(context)
        val parishes = parishTable.parishData
        val headers = parishTable.columns
        if (Environment.MEDIA_MOUNTED != state) {
            Toast.makeText(context, "Media is not mounted.", Toast.LENGTH_SHORT).show()
            return
        } else {
            val printWriter = exportCsv("parish")
            val csvHeader = TextUtils.join(",", headers)
            printWriter!!.println(csvHeader)
            for (p in parishes) {
                val detailsLine = (p.id + ","
                        + p.name.replace(",".toRegex(), ";") + ","
                        + p.country.replace(",".toRegex(), ";") + ","
                        + p.parent.replace(",".toRegex(), ";") + ","
                        + p.mapping.replace(",".toRegex(), ";") + ","
                        + UserTable(context).getUserById(p.addedBy!!)!!.name + ","
                        + p.contactPerson.replace(",".toRegex(), ";") + ","
                        + p.contactPersonPhone.replace(",".toRegex(), ";") + ","
                        + p.comment.replace(",".toRegex(), ";") + ","
                        + p.synced + ","
                        + p.dateAdded)
                printWriter.println(detailsLine)
            }
            printWriter?.close()

        }
    }

    fun exportMapping() {
        val mappingTable = MappingTable(context)
        val mappings = mappingTable.mappingData
        val headers = mappingTable.columns
        if (Environment.MEDIA_MOUNTED != state) {
            Toast.makeText(context, "Media is not mounted.", Toast.LENGTH_SHORT).show()
            return
        } else {
            val printWriter = exportCsv("mapping")
            val csvHeader = TextUtils.join(",", headers)
            printWriter!!.println(csvHeader)
            for (m in mappings) {
                val detailsLine = (m.id + ","
                        + m.mappingName.replace(",".toRegex(), ";") + ","
                        + m.country.replace(",".toRegex(), ";") + ","
                        + m.county.replace(",".toRegex(), ";") + ","
                        + UserTable(context).getUserById(m.addedBy!!)!!.name + ","
                        + m.contactPerson.replace(",".toRegex(), ";") + ","
                        + m.contactPersonPhone.replace(",".toRegex(), ";") + ","
                        + m.comment.replace(",".toRegex(), ";") + ","
                        + m.dateAdded + ","
                        + m.isSynced + ","
                        + m.district.replace(",".toRegex(), ";") + ","
                        + m.subCounty.replace(",".toRegex(), ";"))
                printWriter.println(detailsLine)
            }
            printWriter?.close()

        }

    }

    private fun exportCsv(dataToExport: String): PrintWriter? {
        val exportDir = File(Environment.getExternalStorageDirectory().toString() + "/Download/tremap/")
        if (!exportDir.exists()) {
            exportDir.mkdirs()
        }
        val file: File
        var printWriter: PrintWriter? = null
        file = File(exportDir, "$dataToExport.csv")
        try {
            file.createNewFile()
            printWriter = PrintWriter(FileWriter(file))

        } catch (e: Exception) {
        }

        return printWriter
    }
}
